moulinette/authenticators/__init__.py

```diff
@@ -135,33 +144,62 @@ def _open_sessionfile(self, session_id, mode='r'):
+        to_hash = "{id}:{token}".format(id=session_id, token=session_token)
+        hash_ = hashlib.sha256(to_hash).hexdigest()
         with self._open_sessionfile(session_id, 'w') as f:
-            f.write(s)
+            f.write(hash_)
 
-    def _retrieve_session(self, session_id, session_hash):
+    def _authenticate_session(self, session_id, session_token):
         """Retrieve a session and return its associated password"""
         try:
+            # FIXME : shouldn't we also add a check that this session file
+            # is not too old ? e.g. not older than 24 hours ? idk...
+
             with self._open_sessionfile(session_id, 'r') as f:
-                enc_pwd = f.read()
+                stored_hash = f.read()
         except IOError as e:
             logger.debug("unable to retrieve session", exc_info=1)
             raise MoulinetteError('unable_retrieve_session', exception=e)
         else:
-            gpg = gnupg.GPG()
-            gpg.encoding = 'utf-8'
-
-            decrypted = gpg.decrypt(enc_pwd, passphrase=session_hash)
-            if decrypted.ok is not True:
-                error_message = "unable to decrypt password for the session: %s" % decrypted.status
-                logger.error(error_message)
-                raise MoulinetteError('unable_retrieve_session', exception=error_message)
-            return decrypted.data
+            #
+            # session_id (or just id) : This is unique id for the current session from the user. Not too important 
+            # if this info gets stolen somehow. It is stored in the client's side (browser) using regular cookies.
+            #
+            # session_token (or just token) : This is a secret info, like some sort of ephemeral password, 
+            # used to authenticate the session without the user having to retype the password all the time... 
+            #    - It is generated on our side during the initial auth of the user (which happens with the actual admin password)
+            #    - It is stored on the client's side (browser) using (signed) cookies.
+            #    - We also store it on our side in the form of a hash of {id}:{token} (c.f. _store_session). 
+            #      We could simply store the raw token, but hashing it is an additonal low-cost security layer
+            #      in case this info gets exposed for some reason (e.g. bad file perms for reasons...)
+            #
+            # When the user comes back, we fetch the session_id and session_token from its cookies. Then we
+            # re-hash the {id}:{token} and compare it to the previously stored hash for this session_id ...
+            # It it matches, then the user is authenticated. Otherwise, the token is invalid.
+            #
+            to_hash = "{id}:{token}".format(id=session_id, token=session_token)
+            hash_ = hashlib.sha256(to_hash).hexdigest()
+
+            if hash_ != stored_hash:
```

Comment:
Would it be useful to use `mac.compare_digest()`?

Reply:
I aggree with @randomstuff using `hmac`'s `compare_diggest` would be better https://docs.python.org/3.6/library/hmac.html#hmac.compare_digest
```suggestion
            if not hmac.compare_digest(hash_, stored_hash):
```

Reply:
Yes ideed, that's significantly better, thanks for that ! (Will `import hmac` in another commit)